### PR TITLE
Add meta restoration tests and bug fixes

### LINKED ISSUE #4781: Fix edge cases where an entry is autorestored and deleted in the same TX
Make sure restored -> deleted works and outputs LEDGER_ENTRY_RESTORED, LEDGER_ENTRY_DELETED meta

Make sure we fix the "double restore" bug, where TX 0 restores(e) -> deletes(e) then TX 1 restores(e) in the same ledger

Get the meta right for the TX 0 restore -> delete followed by TX 1 create in the same ledger:

TX 0 meta LEDGER_ENTRY_RESTORE, LEDGER_ENTRY_DELETE
TX 1 meta LEDGER_ENTRY_CREATE

### PR DESCRIPTION
**Pull request by SirTyson**
# Description

Resolves #4781

Adds more restore meta unit tests and fixes a few bugs found from the unit tests. In particular, we had some gaps in coverage where an entry was autorestored and deleted in the same ledger. I've added unit tests for several interesting autorestore scenarios:

1. TX 0 restores and deletes an entry, TX 1 re-creates that same entry in the same ledger
2. TX 0 restores an entry, TX 1 modifies that entry in the same ledger
3. A transaction autorestores and deletes an entry
4. General unit tests around lastModifiedLedgerSeq in restore meta

There was also a bug with back filling restore meta prior to p23, where the same entry may emit restore meta in two different TXs in the same ledger. Finally, there was a bug where the RestoreOp meta would use an outdated lastModifiedLedgerSeq number when restoring an evicted entry in p23.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/transactions/TransactionMeta.cpp [hunk 1]
```diff
@@ -282,13 +302,64 @@ OperationMetaBuilder::setLedgerChanges(AbstractLedgerTxn& opLtx,
...
+    }
+
+    // getRestoredHotArchiveKeys and getRestoredLiveBucketListKeys return all
+    // entries that have been restored this ledger, not just by this op.
+    // However, processOpLedgerEntryChanges expects just the map of restores for
+    // this op. This function only gets called for <p23, so we only have to
+    // worry about the restore op. We look at the TTLs that have been modified
+    // by this op (i.e. restored TTLs) and use that to create an op-specific
+    // subset of the restored key maps.
+    UnorderedMap<LedgerKey, LedgerEntry> opRestoredHotArchiveKeys{};
```
Comment (dmkozh) [inline src/transactions/TransactionMeta.cpp]: There is no hot archive prior to p23, so why do we need this?

File: src/transactions/test/InvokeHostFunctionTests.cpp [hunk 2]
```diff
@@ -3307,11 +3317,81 @@ TEST_CASE_VERSIONS("archival meta", "[tx][soroban][archival]")
...
+                                   LedgerKeySet const& expectedKeys) {
+            auto frame = getMetaForLedger(ledgerSeq);
+            REQUIRE(frame.has_value());
+
+            auto evictedKeys = frame->getEvictedKeys();
+            REQUIRE(evictedKeys.size() == expectedKeys.size());
+
+            for (auto const& key : expectedKeys)
+            {
+                REQUIRE(std::find(evictedKeys.begin(), evictedKeys.end(),
```
Comment (dmkozh) [inline src/transactions/test/InvokeHostFunctionTests.cpp]: nit: Can we sort both arrays and compare them? I imagine this will be really annoying to debug if it fails.

**Comment by SirTyson** [comment]
Comments should be addressed.

**Review by dmkozh — DISMISSED** [review]

File: src/transactions/TransactionMeta.cpp [hunk 3]
```diff
@@ -282,13 +302,52 @@ OperationMetaBuilder::setLedgerChanges(AbstractLedgerTxn& opLtx,
...
+    // by this op (i.e. restored TTLs) and use that to create an op-specific
+    // subset of the restored key maps.
+    UnorderedMap<LedgerKey, LedgerEntry> opRestoredLiveBucketListKeys{};
+    auto allRestoredLiveBucketListKeys = opLtx.getRestoredLiveBucketListKeys();
+    auto opModifiedTTLKeys = opLtx.getAllTTLKeysWithoutSealing();
+    if (mOp.getOperation().body.type() == OperationType::RESTORE_FOOTPRINT)
+    {
+        for (auto const& [key, entry] : allRestoredLiveBucketListKeys)
+        {
+            if (key.type() == CONTRACT_CODE || key.type() == CONTRACT_DATA)
```
Comment (sisuresh) [inline src/transactions/TransactionMeta.cpp]: Nit: You can use `isSorobanEntry(key)` here instead.
